runners/flink/src/main/java/org/apache/beam/runners/flink/translation/wrappers/streaming/ExecutableStageDoFnOperator.java

```diff
@@ -368,6 +386,19 @@ private void prepareStateBackend(K key) {
           // Key for state request is shipped encoded with NESTED context.
           ByteBuffer encodedKey = FlinkKeyUtils.fromEncodedKey(key);
           keyedStateBackend.setCurrentKey(encodedKey);
+          if (keyStateBackendWithKeyGroupInfo != null) {
```

Comment:
Goof to have this extra feedback. It appears these operations are cheap enough to perform on every state access, otherwise they could be gated by a debug flag.

Reply:
Originally, I had this guarded with debug, but considering all the issues we had with this, I went ahead and enforced the check. I've not benchmarked it but this should be extremely  cheap, because the key group index is already computed with `setCurrentKey` and the key group range is static for the operator.

Reply:
So basically, the only cost is the check in the key group range which is `keyGroup >= startKeyGroup && keyGroup <= endKeyGroup;` and the overhead of calling the method; should be negligible.

Reply:
Yes, that's what I found also. You could add a comment as well.